File: frontend/src/services/product.service.ts
import api from "./api";
import { ApiResponse, Category, PageResponse, Product, Review } from "@/types";

export interface ProductFilters {
  keyword?: string;
  category?: string;
  minPrice?: number;
  maxPrice?: number;
  flowerType?: string;
  color?: string;
  inStock?: boolean;
  sort?: string;
  page?: number;
  size?: number;
}

export const productService = {
  async getCategories() {
    const response = await api.get<ApiResponse<Category[]>>("/categories");
    return response.data;
  },

  async getProducts(params?: ProductFilters) {
    const response = await api.get<ApiResponse<PageResponse<Product>>>("/products", { params });
    return response.data;
  },

  async getFeaturedProducts() {
    const response = await api.get<ApiResponse<Product[]>>("/products/featured");
    return response.data;
  },

  async getBestSellers() {
    const response = await api.get<ApiResponse<Product[]>>("/products/best-sellers");
    return response.data;
  },

  async getNewArrivals() {
    const response = await api.get<ApiResponse<Product[]>>("/products/new-arrivals");
    return response.data;
  },

  async getProductBySlug(slug: string) {
    const response = await api.get<ApiResponse<Product>>(`/products/${slug}`);
    return response.data;
  },

  async getRelatedProducts(id: number) {
    const response = await api.get<ApiResponse<Product[]>>(`/products/${id}/related`);
    return response.data;
  },

  async getProductReviews(productId: number, page = 0, size = 10) {
    const response = await api.get<ApiResponse<PageResponse<Review>>>(`/products/${productId}/reviews`, {
      params: { page, size },
    });
    return response.data;
  },
};

File: backend/src/main/java/com/flowerstore/service/ProductService.java
package com.flowerstore.service;

import com.flowerstore.dto.request.ProductRequest;
import com.flowerstore.dto.response.PageResponse;
import com.flowerstore.dto.response.ProductImageResponse;
import com.flowerstore.dto.response.ProductResponse;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.math.BigDecimal;
import java.util.List;

public interface ProductService {
    PageResponse<ProductResponse> getProducts(
            String keyword,
            String categorySlug,
            BigDecimal minPrice,
            BigDecimal maxPrice,
            String flowerType,
            String color,
            Boolean inStock,
            String sort,
            Pageable pageable
    );

    ProductResponse getProductBySlug(String slug);
    ProductResponse getProductById(Long id);
    List<ProductResponse> getFeaturedProducts();
    List<ProductResponse> getBestSellers();
    List<ProductResponse> getNewArrivals();
    List<ProductResponse> getRelatedProducts(Long id);

    // Admin APIs
    PageResponse<ProductResponse> getAdminProducts(String keyword, Pageable pageable);
    ProductResponse createProduct(ProductRequest request);
    ProductResponse updateProduct(Long id, ProductRequest request);
    void softDeleteProduct(Long id);
    ProductResponse restoreProduct(Long id);
    ProductResponse toggleProductStatus(Long id, Boolean isActive);
    ProductImageResponse addProductImage(Long productId, MultipartFile file, Boolean isPrimary);
    void deleteProductImage(Long productId, Long imageId);
}
